content/blog/2019-11-13-llvm-function-inlining/index.md

```diff
@@ -0,0 +1,144 @@
+and the instruction pointer will need to be moved to the beginning of that
+function—which may or may not be in the instruction cache—among other
+function-initialization tasks. Also, some of the state of the calling function
+(e.g., caller-saved registers) will need to be stored in preparation for the
+call, such that it will not be overwritten by the called function. Separating
+out functions also gives the compiler less information for optimizing the
+program, as, when looking at a function, the compiler cannot be sure of the
+context that it will be running in. As such, it is often beneficial for
+compilers to be able to get rid of function calls.
+
+Function inlining is a classic compiler optimization that does exactly this; it
+replaces calls in code with the body of the called function in order to remove
+the overhead of making the call at runtime. In this project we implemented and
+evaluated the effects of a function inlining pass which runs on programs in the
+LLVM intermediate representation (IR).
+
+## Design
+
+To properly inline a function call without breaking the program, there are
+several considerations that must be made. As a first approximation, the call
+instruction is erased, and all of the instructions in the function that was
+called are put in its place. Any arguments or return values in the call must
+then be handled. Uses of the arguments in the inlined function now need to refer
+to the actual values that were provided in the call instruction. Due to the
+possibility of multiple return instructions in the function, handling the data
+flow and control flow for exiting the inlined function is a bit more
+complicated. All return instructions are replaced with unconditional branches to
+the block which immediately followed the call in the parent function. A phi node
+is inserted into that block to coalesce the values from all of the return
+instructions into a single place, which then can be used whereever the returned
+value was used in the parent function. The LLVM API provides the function
+`llvm::InlineFunction` which handles completing these operations. In general,
+you would also need to handle naming collisions between variables in the caller
+and callee functions. However, because the LLVM IR is in static single
+assignment (SSA) form, and data dependencies are represented explicitly through
+pointers, this is not a concern.
+
+With a method for inlining in place, the next step is to determine when we want
+to inline calls. Some function calls can be immediately eliminated from
+consideration because they would be impossible to inline correctly. For example,
+functions not defined locally cannot be inlined, as their implementations are
+unknown. Also, any functions that include indirect branch instructions cannot be
+inlined as doing so would modify the functions' memory locations, and as such
+could cause the indirect branches to go to unexpected positions in the program.
```

Comment:
Are you referring to the `indirectbr` instruction? Typically, these instructions get their addresses using `blockaddress` instructions to get the address of a label, so they're usually "function relative." So it would be possible with a little more effort to classify most of these indirect jumps as safely inlinable.